### PR TITLE
Added basic test cases with renode

### PR DESCRIPTION
**Pull request by danielinux**
- Added renode test script
- Added `renode-factory-*` test targets and `renode-factory-all` to test secure boot on emulated STM32F40x
- Update tests not yet possible (looks like flash write is not yet suppored on this platform)

### CODE REVIEW

**Comment by danielinux** [comment]
`.config` that can be used with `make renode-factory-all` so far:
 - No `config` : defaults to stm32F407-discovery
 - `config/examples/stm32f7.config`
 - `config/examples/hifive1.config`
 - `config/examples/nrf52840.config`

File: tools/scripts/renode-test-all.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+set -x
```
Comment (elms) [inline tools/scripts/renode-test-all.sh:2]: Maybe `set -e` in to stop on first error. May require setting a trap to cleanup. Or you could break and set an exit code to call at the end.

File: tools/test-renode.mk [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+endif
+ifeq ($(HASH),SHA3)
+  SIGN_ARGS+= --sha3
+endif
+
+# Testbed actions
+#
+#
+renode-on: FORCE
+	@rm -f /tmp/wolfboot.uart
```
Comment (elms) [inline tools/test-renode.mk:82]: `${Q}` over `@` to allow showing commands.

File: tools/test-renode.mk [hunk 3]
```diff
@@ -0,0 +1,139 @@
+TEST_UPDATE_VERSION?=2
+WOLFBOOT_VERSION?=0
+RENODE_UART=/tmp/wolfboot.uart
```
Comment (elms) [inline tools/test-renode.mk:3]: Would it make sense to use `$TMP` isntead of `/tmp` or maybe even `mktemp` (may be less compatible)

File: tools/test-renode.mk [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+RENODE_PORT=55155
+RENODE_OPTIONS=--pid-file=/tmp/renode.pid
```
Comment (elms) [inline tools/test-renode.mk:5]: use variable for `/tmp/renode.pid` so other uses can refer to the variable

File: tools/test-renode.mk [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+  SIGN_ARGS+= --sha3
+endif
+
+# Testbed actions
+#
+#
+renode-on: FORCE
+	@rm -f /tmp/wolfboot.uart
+	@renode $(RENODE_OPTIONS) $(RENODE_CONFIG) 2>&1 >/tmp/renode.log &
+	@while ! (test -e /tmp/wolfboot.uart); do sleep .1; done
```
Comment (elms) [inline tools/test-renode.mk:84]: replace `/tmp/wolboot.uart` with `RENODE_UART` variable in all the targets

**Review by elms — CHANGES_REQUESTED** [review]
This is really awesome. Very exciting to see this work. Just a few little things...and I want to test a little more.

I was able to test by running(Except riscv, I'm having issues with compiler setup)

```
make keytools
make tools/test-expect-version/test-expect-version
./tools/scripts/renode-test-all.sh
```

A few suggestions for the makefile:
 * targets depend on `test-expected-version`
 * Maybe out of scope.... Default to use build and use C keytools or detect if python package needs to be installed (`python -c "import wolfcrypt" &> /dev/null`) and suggest the command `pip install wolfcrypt`

**Comment by danielinux** [comment]
Thanks @elms
fixed most points.

I don't think that defaulting to C tools should be part of this PR. 
Other tests use the same logic (if a compiled tool exist use it, otherwise fallback to the default .py), if we change this here we should align `tools/test*` too.

File: tools/scripts/renode-test-all.sh (hunk 1 above)
Comment (danielinux) [reply tools/scripts/renode-test-all.sh]: Removed `-x`. Disagree on `-e`, but I've added exit points on `make` invocations.

**Review by elms — APPROVED** [review]
LGTM.

I was still unable to test with riscv, but looks like it's an issue with my toolchain. Would be good to document the requirements and maybe even a docker image we can leverage for it.

**Pull request merged**